### DISCUSSION TITLE
`autoclear_cookies` doesn't work in React

### DISCUSSION THREAD

**Discussion by KitaLance** [orestbida/cookieconsent#550] [Q&A] [answered]
I'm trying to implement this plugin in my React project and I found the `autoclear_cookies` setting doesn't work, here's my setting for plugin config:

```   
        ...
        settings_modal: {
        title: 'Cookie Settings',
        save_settings_btn: 'Save Settings',
        accept_all_btn: 'Accept All',
        reject_all_btn: 'Reject All',
        close_btn_label: 'Close',
        cookie_table_headers: [
          { col1: 'Name' },
          { col2: 'Domain' },
          { col3: 'Expiration' },
          { col4: 'Description' },
        ],
        blocks: [
          {
            description: '...',
          },
          {
            title: 'Strictly Necessary',
            description: '...',
            toggle: {
              value: 'necessary',
              enabled: true,
              readonly: true,
            },
          },
          {
            title: 'Performance Cookies',
            description:  '...',
            toggle: {
              value: 'analytics',
              enabled: false,
              readonly: false,
            },
            cookie_table: [
              {
                col1: 'muc_ads',
                col2: 't.co',
                col3: '1 year',
                col4: 'description...',
              },
            ],
          },
          },
        ],
      },
```

I didn't set `onAccept` or `onChange` since it looks unnecessary in the doc if I set `autoclear_cookies`. Do I need to add `window.CC.eraseCookies()` to clear in `onChange` manually?
How the items in `cookie_table_headers` really work? Seems that `Domain` is just some value displayed on the table, and only `Name` really works? cmiiw
btw it's an amazing plugin, thanks for your great work

**Comment by KitaLance** [comment]
Updated: I tried v3 and seems that `autoClear` doesn't work either

**Reply by dombyte** [reply]
In v3 autoClear has renamed to autoClearCookies.

**Comment by dombyte** [comment]
Can you give me your full Configuration, maybe I can help.

**Reply by KitaLance** [reply]
Here's my config:
```
const pluginConfig: CookieConsentConfig = {
  guiOptions: {
    consentModal: {
      layout: 'box inline',
      position: 'bottom right',
      equalWeightButtons: true,
      flipButtons: false
    },
    preferencesModal: {
      layout: 'box',
      position: 'left',
      equalWeightButtons: true,
      flipButtons: false
    }
  },

  onFirstConsent: function () {
  },

  onConsent: function () {
  },

  onChange: function () {
  },

  categories: {
    necessary: {
      readOnly: true,
      enabled: true
    },
    performance: {
      autoClear: {
        cookies: [
          {
            name: /^(_ga|_gid)/
          },
          {
            name: 'muc_ads'
          }
        ],
      }
    },
  },

  language: {
    default: 'en',

    translations: {
      en: {
        consentModal: {
          title: 'Your privacy',
          description: '...',
          acceptAllBtn: 'Accept All',
          acceptNecessaryBtn: 'Necessary Only',
          showPreferencesBtn: 'Customize Settings'
        },
        preferencesModal: {
          title: 'Cookie Setting',
          acceptAllBtn: 'Accept All',
          acceptNecessaryBtn: 'Necessary Only',
          savePreferencesBtn: 'Save Preferences',
          closeIconLabel: 'Close',
          sections: [
            {
              title: 'Cookie Usage',
              description: '...',
            },
            {
              title: 'Strictly Necessary',
              description: '...',
              linkedCategory: 'necessary',
            },
            {
              title: 'Performance Cookies',
              description: '...',
              linkedCategory: 'performance',
            }
          ]
        }
      }
    }
  }
}
```
and after toggling `Performance Cookies` off the cookies named `muc_ads` or started with `_ga` can still be found in `Chrome - Application - Cookie`
<img width="361" alt="Screen Shot 2023-07-14 at 14 24 14" src="https://github.com/orestbida/cookieconsent/assets/40763911/c3de9825-93eb-491d-b746-da5fd9f7e59e">

**Reply by dombyte** [reply]
I may found something. You call the autoclear option in the performance category, but it must set in:
```js
cc.run({
    // ...
    autoclear_cookies: true,
    // ...
});
```
Docs: https://github.com/orestbida/cookieconsent/tree/master#how-to-clear-cookies

I also See that you haven't set a cookie for the plugin, im not sure if its necessary.
I used the Docs for reference and the Demos:
https://stackblitz.com/edit/nextjs-euxk9k?file=components%2FCookieConsentConfig.js
https://github.com/orestbida/cookieconsent/blob/master/demo/demo_gtm/cookieconsent-init.js

For me it helps when using the Demos or an example Code as an starting point and then modifying it step for step for my needs.
Especially for your needs, because I don't know anything about working with React there for the [Demo](https://stackblitz.com/edit/nextjs-euxk9k?file=package.json) React Project is a great help. 

Let me know if this helps.

Maybe this config works:
```js
const pluginConfig: CookieConsentConfig = {
  autoclear_cookies: true,
  cookie_name: 'cc_cookie',


  guiOptions: {
    consentModal: {
      layout: 'box inline',
      position: 'bottom right',
      equalWeightButtons: true,
      flipButtons: false
    },
    preferencesModal: {
      layout: 'box',
      position: 'left',
      equalWeightButtons: true,
      flipButtons: false
    }
  },

  onFirstConsent: function () {
  },

  onConsent: function () {
  },

  onChange: function () {
  },

  categories: {
    necessary: {
      readOnly: true,
      enabled: true
    },
    performance: {
        cookies: [
          {
            name: /^(_ga|_gid)/
          },
          {
            name: 'muc_ads'
          }
        ],
      }
    },
  },

  language: {
    default: 'en',

    translations: {
      en: {
        consentModal: {
          title: 'Your privacy',
          description: '...',
          acceptAllBtn: 'Accept All',
          acceptNecessaryBtn: 'Necessary Only',
          showPreferencesBtn: 'Customize Settings'
        },
        preferencesModal: {
          title: 'Cookie Setting',
          acceptAllBtn: 'Accept All',
          acceptNecessaryBtn: 'Necessary Only',
          savePreferencesBtn: 'Save Preferences',
          closeIconLabel: 'Close',
          sections: [
            {
              title: 'Cookie Usage',
              description: '...',
            },
            {
              title: 'Strictly Necessary',
              description: '...',
              linkedCategory: 'necessary',
            },
            {
              title: 'Performance Cookies',
              description: '...',
              linkedCategory: 'performance',
            }
          ]
        }
      }
    }
  }
}
```

**Reply by KitaLance** [reply]
thanks for the response
in v3 `autoClearCookies` is set to `true` by default, and it doesn't work when I manually set it to `true` either
I develop my app based on this new v3 demo: https://stackblitz.com/edit/nextjs-epd84e?file=components%2FCookieConsentComponent.tsx other parts work fine so far except for this clear cookie.

**Reply by dombyte** [reply]
Oh so you use v3? Then I made a mistake. Its `autoClearCookies: true,` not `autoclear_cookies: true,`
I saw in the FAQ section this: 
```

> How to create custom cookie tables

- Cookie tables are defined by you, that is you choose how many columns and what their naming will be
- Make sure that the first column of the table contains the name of the cookie for autoclear_cookie to work properly
```
So maybe take a look into the [FAQ](https://github.com/orestbida/cookieconsent#faq) and look unter the _How to create custom cookie tables_ Section.
I think I also had this problem in the beginning and could solve it by adding a Cookie Table.

My current project uses v3, its plain JS but maybe for some inspiration, because there it works.
Config: https://codefile.io/f/UeUhn5ibjy
EN Translation: https://codefile.io/f/4LTAd3nx0G

**Reply by KitaLance** [reply]
I think I follow the FAQ or tutorial strictly and still, the auto clear doesn't work
But I'm just wondering how it looks like if it works? I wiped the cookies accessed by `document.cookie` in `onchange`, the cookies are cleared at first, but came back after refreshing. Seems that this plugin cannot block the cookies in `autoClear` table?

**Reply by dombyte** [reply]
Hmm, it's sounds like the cookies aren't set by the Plug-in. Are the cookies set by an external script or so? 
Because then it would make sense, you can't block cookies which are not set by the plug-in, therefore you need to block the script which set the cookies. 
Or do I completely misunderstood you and your problem.

**Reply by KitaLance** [reply]
yeah it's a third party script issue, this plugin works fine, amazing
thanks for your help :)

**Comment by orestbida** [comment, marked as answer]
@KitaLance the autoclear function works fine on my end; you need to block the scripts that set these cookies, and only execute them if consent is given; otherwise cookies will always be set after a page refresh.

Since you're on React, you will need to rely on the `onAccept`/`onChange` callbacks to enable/disable the scripts.

**Reply by KitaLance** [reply]
(edited) working now. thanks for your great work!
Solution in short: For React, need to remove the third party script/iframe node manually in `onConsent` / `onChange`
